fix(docs-infra): fix redirecting `rc.angular.io` to `angular.io` when no active RC

gkalpak opened the pull request:
Currently there is an issue with redirecting `rc.angular.io` to `angular.io` when there is no active RC. If a user has visited `rc.angular.io` before and has a ServiceWorker registered for that subdomain, they will never "see" the redirect to `angular.io`.

This PR fixes the problem by doing an additional deployment from the stable branch to the `rc-angular-io-site` Firebase site when there is no active RC. This additional deployment will ensure that:
1. Users will be temporarily redirected from `rc.angular.io` to `angular.io`.
2. Users with a registered ServiceWorker (who don't see the redirect) will have their ServiceWorker unregistered on the next visit.
3. The content on both sites is identical.

See #39760 for more details on the problem and the solution.

NOTE:
As mentioned in #39760, for this fix to take affect, we need to remove the redirect from `rc.angular.io` to `angular.io` in the Firebase console for site `rc-angular-io-site`.

Fixes #39760.

##
This PR also contains some other fixes and refactorings. See individual commits for more details.

mary-poppins commented:
You can preview f96d768 at https://pr39853-f96d768.ngbuilds.io/.

mary-poppins commented:
You can preview 4256ba5 at https://pr39853-4256ba5.ngbuilds.io/.

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -131,47 +132,58 @@ function computeDeploymentsInfo(
+  // Determine if there is an active RC version by checking whether the most recent minor branch is
+  // the stable branch or not.
+  const mostRecentMinorBranch = getMostRecentMinorBranch();
+  const rcBranch = (mostRecentMinorBranch !== stableBranch) ? mostRecentMinorBranch : null;
```
So we are saying that the most recent minor branch that is not the stable branch must be an RC?
Is it worth capturing this logic into a separate helper function. I feel this computation would be needed elsewhere too?

gkalpak replied:
That's the logic. I didn't find it being useful elsewhere. Did you have something specific in mind?

petebacondarwin replied:
I thought there might be other decision points for RC releases but I guess they are all handled here...

petebacondarwin commented on aio/scripts/deploy-to-firebase.spec.js:
```diff
@@ -302,7 +316,7 @@ describe('deploy-to-firebase:', () => {
         reason:
             'Skipping deploy of branch "2.1.x" to Firebase.\n' +
-            'There is a more recent branch with the same major version: "2.4.x"',
+            'This branch has an equal or higher major version than the stable branch ("2.0.x").',
```
I think the test description needs tweaking here. Perhaps...

```
it('rc - skip deploy - major same as stable, branch not the most recent minor', () => {
```

gkalpak replied:
How so? The current description is accurate afaict. Isn't "branch not the most recent minor" the same as "minor not latest"?

petebacondarwin replied:
`latest` in my mind (while reading this) is more like "stable" or "master". But if you agree that we are talking about the latest minor "for the given major that we are checking" it is equivalent. It just took me longer to grok. So I thought being more explicit would help.

gkalpak replied:
That's how we use "latest" throughout these tests. Happy switch to "most recent".

gkalpak replied:
Actually, I switched to `highest` to be more consistent with other references to `higher/-est`/`lower/-est` in the test descriptions.

petebacondarwin commented on aio/scripts/deploy-to-firebase.spec.js:
```diff
@@ -208,7 +215,7 @@ describe('deploy-to-firebase:', () => {
         reason:
             'Skipping deploy of branch "2.1.x" to Firebase.\n' +
-            'There is a more recent branch with the same major version: "2.4.x"',
+            'This branch has an equal or higher major version than the stable branch ("2.2.x").',
```
I feel that the old reason for skipping is more appropriate here... Is there any way we could change the order of the checks to return to that reason message?

gkalpak replied:
Yeah, I would prefer the old reason to be the first to check as well. I opted for doing the "same or higher major" check first because it is cheaper (we don't have to fetch remote refs).

Thinking about it again, I don't think it matters much, because we rarely push to branches other than master/RC/stable. I'll change the order of checks.

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -131,47 +132,58 @@ function computeDeploymentsInfo(
+    return [
+      skipDeployment(
+          `Skipping deploy of branch "${currentBranch}" to Firebase.\n` +
+          'This branch has an equal or higher major version than the stable branch ' +
+          `("${stableBranch}").`),
```
```suggestion
          'This branch has an equal or higher major version than the stable branch ' +
          `("${stableBranch}") and is not the latest minor branch.`),
```

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -263,6 +285,21 @@ function getLatestCommit(branchName, remote = undefined) {
+function removeServiceWorker() {
+  // Rename the SW manifest (`ngsw.json`). This will cause the ServiceWorker to unregister itself.
+  // See https://angular.io/guide/service-worker-devops#fail-safe.
+  mv('dist/ngsw.json', 'dist/ngsw.json.bck');
```
OOC why `bck`?

gkalpak replied:
Because I believe there is a 62% chance in the future we will require to undo these modifications once deployed (for example, to deploy somewhere else). So, I am essentially keeping a backup to be able to restore the file later. In fact, there is an 11.2% chance a reviewer will request the clean-up be done in this PR 🤯

petebacondarwin replied:
And `bck` rather than `bak`?

gkalpak replied:
Ah, good question :sweat_smile:
I didn't think about it. I guess, that is a personal "convention" (because many programs create `.bak` files, I use `.bck` to avoid messing with other programs).

I'll change it to `.bak` to avoid confusion.

gkalpak replied:
Fixed: https://github.com/angular/angular/pull/39853/commits/30f8665db6a41491226e4407db9ce4fa269c7459

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -263,6 +285,21 @@ function getLatestCommit(branchName, remote = undefined) {
+  // not contain a dot in their last path segment) to `angular.io`.
+  // See https://firebase.google.com/docs/hosting/full-config#redirects.
+  const redirectRule =
+      '{"type": 302, "regex": "^(.*/[^./]*)$", "destination": "https://angular.io:1"}';
```
Have you tested that `:1` substitutes the group? I couldn't find this in the Firebase docs.

petebacondarwin replied:
What happens if there is a fragment or query parameter with a `.` or `/` in it? E.g. `https://rc.angular.io/foo/bar?query=foo.bar`.

gkalpak replied:
> Have you tested that `:1` substitutes the group? I couldn't find this in the Firebase docs.

Yes, I have tested it. It's legit 😛
It is mentioned in the docs [here][1] (you have to expand the `Capture URL segments when using RE2 regular expressions` section).

##
> What happens if there is a fragment or query parameter with a `.` or `/` in it? E.g. `https://rc.angular.io/foo/bar?query=foo.bar`

The server doesn't see the hash fragment, but good point on the query param. I'll change the regex:

```diff
-^(.*/[^./]*)$
+^(.*/[^./?]*)(?:\?|$)
```

[1]: https://firebase.google.com/docs/hosting/full-config#redirects

gkalpak replied:
Actually, I tried it and it turns out that the existing regex works. With a more careful reading, I found out in the [hosting config docs][1] that only the **URL path** is taken into account for redirects:

> In each rule, specify a URL pattern that, if matched to the request URL path, triggers Hosting to respond with a redirect to the specified destination URL.

So, I am going to leave the regex as `^(.*/[^./]*)$`. (Simpler is better :grin:)

[1]: https://firebase.google.com/docs/hosting/full-config#redirects

petebacondarwin commented on aio/scripts/deploy-to-firebase.js:
```diff
@@ -271,6 +308,41 @@ function skipDeployment(reason) {
   return {reason, skipped: true};
 }
 
+function testNoActiveRcDeployment({deployedUrl}) {
```
Is there any way to run this test to make sure everything really works before we go live with it?

gkalpak replied:
I have tried deploying to a project of mine and then tested with this function and it worked 🤷‍♂️

petebacondarwin reviewed (approved):
I love love love the separate commits! This PR would have been a nightmare to review if it was all in one commit. Thank you for that.

A few minor comments but all in all LGTM.

Small typo (+query) for this commit message:

In `fix(docs-infra): do not deploy as `archive` when major is not lower than stable`

> Typically, there will only be at most one minor branch ~~most~~more recent than
the stable one (and that branch will be the RC branch).

(or the master branch?)

gkalpak commented:
Thx for the speedy review, @petebacondarwin 🚀
I've addressed your comments. PTAL

> Small typo (+query) for this commit message:
>
>> In fix(docs-infra): do not deploy as archive when major is not lower than stable
>>
>> Typically, there will only be at most one minor branch ~~most~~more recent than
>> the stable one (and that branch will be the RC branch).
>
> (or the master branch?)

By "minor branch", I meant branches with names of the form `X.Y.*` (which correspond to the `Y` minor of the `X` major).

mary-poppins commented:
You can preview cff57f5 at https://pr39853-cff57f5.ngbuilds.io/.

gkalpak commented:
Thx for the speedy review, @petebacondarwin 🚀
I've addressed your comments (added two `fixup` commits and a regular one at the end to address [this comment](https://github.com/angular/angular/pull/39853#discussion_r531730317)). PTAL

> Small typo (+query) for this commit message:
>
>> In fix(docs-infra): do not deploy as archive when major is not lower than stable
>>
>> Typically, there will only be at most one minor branch ~~most~~more recent than
>> the stable one (and that branch will be the RC branch).
>
> (or the master branch?)

By "minor branch", I meant branches with names of the form `X.Y.*` (which correspond to the `Y` minor of the `X` major).

mary-poppins commented:
You can preview 0f45047 at https://pr39853-0f45047.ngbuilds.io/.

mary-poppins commented:
You can preview 44c28d2 at https://pr39853-44c28d2.ngbuilds.io/.

gkalpak committed:
refactor(docs-infra): decouple deploying from other operations in `deploy-to-firebase.js`

Previously, the `deploy()` function in `deploy-to-firebase.js` would
also perform other operations (beyond deploying), such as building the
app, checking the generated payload size, testing the PWA score of the
deployed app.

This commit decouples these operations, so that deploying can be
performed independently.

NOTE:
In a subsequent commit, this will be leveraged fix an issue with
`rc.angular.io` redirects in the presence of a ServiceWorker by
deploying the same artifacts to multiple Firebase projects/sites.
See #39760 for more details.

gkalpak committed:
test(docs-infra): avoid unnecessary re-computations in `deploy-to-firebase.spec.js`

Previously, the latest commit for branch may be computed multiple times
in the `deploy-to-firebase.js` tests.

This commit avoids the unnecessary re-computations by computing the
latest commits for the necessary branches at the beginning and using the
computed values throughout the tests.

gkalpak committed:
refactor(docs-infra): support multiple deployments per run

Previously, the `deploy-to-firebase.js` script would only perform one
deployment operation on each run.

This commit adds support for performing multiple deployment operations.

NOTE:
In a subsequent commit, this will be leveraged fix an issue with
`rc.angular.io` redirects in the presence of a ServiceWorker by
deploying the same artifacts to multiple Firebase projects/sites.
See #39760 for more details.

gkalpak committed:
fix(docs-infra): do not deploy as `archive` when major is not lower than stable

Previously, a branch would be deployed as `archive` even if it had a
major version that was equal/higher than that of the stable branch (as
long as it was not the RC branch - i.e. not the most recent minor
branch). For example, with `11.0.x` as the stable branch  and `12.0.x`
as the RC branch, `11.1.x` would be deployed as archive.

Theoretically, we should never find ourselves in such a situation.
Typically, there will only be at most one minor branch most recent than
the stable one (and that branch will be the RC branch). However, it
is possible under unusual circumstances.

This commit adds additional checks to guard against this problem. It
also refactors the code in preparation of fixing an issue with
`rc.angular.io` redirects in the presence of a ServiceWorker, which will
require identifying whether there is an active RC version or not.
See #39760 for more details.

gkalpak committed:
fixup! fix(docs-infra): do not deploy as `archive` when major is not lower than stable

Address review comment:
https://github.com/angular/angular/pull/39853#discussion_r531540007

gkalpak committed:
fix(docs-infra): fix redirecting `rc.angular.io` to `angular.io` when no active RC

Currently there is an issue with redirecting `rc.angular.io` to
`angular.io` when there is no active RC. If a user has visited
`rc.angular.io` before and has a ServiceWorker registered for that
subdomain, they will never "see" the redirect to `angular.io`.

This commit fixes the problem by doing an additional deployment from the
stable branch to the `rc-angular-io-site` Firebase site when there is no
active RC. This additional deployment will ensure that:
1. Users will be temporarily redirected from `rc.angular.io` to
   `angular.io`.
2. Users with a registered ServiceWorker (who don't see the redirect)
   will have their ServiceWorker unregistered on the next visit.
3. The content on both sites is identical.

See #39760 for more details on the problem and the solution.

NOTE:
As mentioned in #39760, for this fix to take affect, we need to remove
the redirect from `rc.angular.io` to `angular.io` in the Firebase
console for site `rc-angular-io-site`.

Fixes #39760

gkalpak committed:
fixup! fix(docs-infra): do not deploy as `archive` when major is not lower than stable

Address review comment:
https://github.com/angular/angular/pull/39853#discussion_r531540621

gkalpak committed:
test(docs-infra): reduce ambiguity in `deploy-to-firebase.js` test descriptions

Previously, test descriptions used `latest` to refer to the most
recent/highest version. This was ambiguous, because `latest` can also
refer to the stable version of a package (e.g. see `@latest` npm tag).

This commit replaces `latest` with `highest` (or `highest for major`) to
reduce ambiguity.

Discussed in:
https://github.com/angular/angular/pull/39853#discussion_r531730317

gkalpak committed:
fixup! fix(docs-infra): fix redirecting `rc.angular.io` to `angular.io` when no active RC

mary-poppins commented:
You can preview 023529f at https://pr39853-023529f.ngbuilds.io/.

IgorMinar reviewed (approved):
LGTM
